src/librustc_privacy/lib.rs

```diff
@@ -522,8 +522,46 @@ impl<'a, 'tcx> Visitor<'tcx> for EmbargoVisitor<'a, 'tcx> {
             hir::ItemKind::ExternCrate(..) => {}
             // All nested items are checked by `visit_item`.
             hir::ItemKind::Mod(..) => {}
-            // Re-exports are handled in `visit_mod`.
-            hir::ItemKind::Use(..) => {}
+            // Re-exports are handled in `visit_mod`. However, in order to avoid looping over
+            // all of the items of a mod in `visit_mod` looking for use statements, we handle
+            // making sure that intermediate use statements have their visibilities updated here.
+            hir::ItemKind::Use(ref path, _) => {
+                if item_level.is_some() {
+                    // If this is a use of another use of an item then we need
+                    // to update the visibility of the intermediate use so that it isn't linted
+                    // by `unreachable_pub`.
+                    //
+                    // This isn't trivial as `path.def` has the `DefId` of the eventual target
+                    // of the use statement not of the next intermediate use statement.
+                    //
+                    // To do this, consider the last two segments of the path to our intermediate
+                    // use statement. We expect the penultimate segment to be a module and the
+                    // last segment to be the name of the item we are exporting. We can then
+                    // look at the items contained in the module for the use statement with that
+                    // name and update that item's visibility.
+                    let segments = path.segments.as_ref();
+                    if let Some(&[ref module, ref name]) = segments.rchunks_exact(2).next() {
```

Comment:
Nit: looks like `Some([module, name])` would work here due to default binding modes.

Reply:
(It's `segment`, not `name` though.)

Reply:
Fixed.